codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java

```diff
@@ -0,0 +1,203 @@
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+/**
+ * A {@link StreamByteDistributor} that ignores stream priority and uniformly allocates bytes to all
+ * streams. This class uses a minimum chunk size that will be allocated to each stream. While
+ * fewer streams may be written to in each call to {@link #distribute(int, Writer)}, doing this
+ * should improve the goodput on each written stream.
+ */
+public final class UniformStreamByteDistributor implements StreamByteDistributor {
+    static final int DEFAULT_MIN_ALLOCATION_CHUNK = 1024;
+
+    private final Http2Connection.PropertyKey stateKey;
+    private final Deque<State> queue = new ArrayDeque<State>();
+    private final Deque<State> emptyFrameQueue = new ArrayDeque<State>();
+
+    /**
+     * The minimum number of bytes that we will attempt to allocate to a stream. This is to
+     * help improve goodput on a per-stream basis.
+     */
+    private int minAllocationChunk = DEFAULT_MIN_ALLOCATION_CHUNK;
+    private long totalStreamableBytes;
+
+    public UniformStreamByteDistributor(Http2Connection connection) {
+        checkNotNull(connection, "connection");
+
+        // Add a state for the connection.
+        stateKey = connection.newKey();
+        Http2Stream connectionStream = connection.connectionStream();
+        connectionStream.setProperty(stateKey, new State(connectionStream));
+
+        // Register for notification of new streams.
+        connection.addListener(new Http2ConnectionAdapter() {
+            @Override
+            public void onStreamAdded(Http2Stream stream) {
+                stream.setProperty(stateKey, new State(stream));
+            }
+
+            @Override
+            public void onStreamClosed(Http2Stream stream) {
+                state(stream).close();
+            }
+        });
+    }
+
+    /**
+     * Sets the minimum allocation chunk that will be allocated to each stream. Defaults to 1KiB.
+     *
+     * @param minAllocationChunk the minimum number of bytes that will be allocated to each stream.
+     * Must be > 0.
+     */
+    public void minAllocationChunk(int minAllocationChunk) {
+        if (minAllocationChunk <= 0) {
+            throw new IllegalArgumentException("minAllocationChunk must be > 0");
+        }
+        this.minAllocationChunk = minAllocationChunk;
+    }
+
+    @Override
+    public void updateStreamableBytes(StreamState streamState) {
+        State state = state(streamState.stream());
+        state.updateStreamableBytes(streamState.streamableBytes(), streamState.hasFrame());
+    }
+
+    @Override
+    public boolean distribute(int maxBytes, Writer writer) throws Http2Exception {
+        checkNotNull(writer, "writer");
+
+        // First, write out any empty frames.
+        while (!emptyFrameQueue.isEmpty()) {
+            State state = emptyFrameQueue.remove();
```

Comment:
> Just to be clear, let me explain the performance advantage of the 2 queue approach ...

Yip benefits are understood :) I didn't notice that while iterating the `emptyFrameQueue` you were checking `state.streamableBytes > 0` and then appending to `queue` instead of writing immediately (did this change recently?).

> The additional memory will be minimal as we don't expect a lot of zero-length frames to be added

Aren't header frames zero-length?  In some cases (when headers and data are written at same time) this may mean we are actually doing 2 iterations over collections of size `n`.

-> talked offline with @nmittler - we should not put headers through the flow controller.